Implement flow mode

SarahAlidoost opened the pull request:
closes #4 

:red_circle: This branch is based on `fix_5` branch. First PR #19 should be merged.

fnattino commented on R/stroke.R:
```diff
@@ -150,6 +171,12 @@ best_link <- function(nodes, segments, links, angle_threshold = 0) {
     best_link <- get_best_link(angles, linked_segs, angle_threshold_rad)
+
+    # if flow_mode, we choose the link on the same edge
+    if (flow_mode) {
+      best_link <- get_link_on_same_edge(linked_segs, edge_ids, edge_id)
+    }
```
Nice way to find the continuation link on the same edge! But what happens if none of the linked segments falls on the same edge? If I understand it correctly,  `get_link_on_same_edge` would then return an empty array, the assignment on line 180 would be skipped, and the  `best_links` element would remain NA. 

I think this is not what should happen here: when we have no more continuation links on the same edge (i.e. when we reach the end of an edge) we actually have to look for the best-angle continuation of the segment. So I wonder whether  one should maybe swap the order of the last few lines? Something like:
1. Call `get_linked_segments` to find out which  continuation links we have;
2. If `flow_mode`, call `get_link_on_same_edge` to find out whether we have a candidate `best_link`
3. If `(lenght(best_link) == 0` (and if not `flow_mode` ),  call `get_linked_nodes` , `interior_angle`, and `get_best_link` to find out the best possible continuation of the segment. 
4.  Only if `(lenght(best_link) > 0` , proceed to the assignment, leave the best link as NA otherwise. 

What do you think? I have to say I haven't tried running these steps in this order, so I am not 100% this will work. Also I am curious if you have more insight by having actually run the tests for `flow_mode = TRUE` (which seems to pass!)..

SarahAlidoost replied:
Correct, I fixed the order of the operations and the tests are passed.

fnattino reviewed (commented):
Thanks for the speedy implementation, amazing! I have basically only one question/comment -  I feel we are very close in getting this working already!

SarahAlidoost commented on R/stroke.R:
```diff
@@ -129,19 +146,23 @@ get_linked_nodes <- function(node_id, segment_id, segments) {
 #' @noRd
-best_link <- function(nodes, segments, links, angle_threshold = 0) {
+best_link <- function(
+  nodes, segments, links, edge_ids, angle_threshold = 0, flow_mode
```
I didnot set `flow_mode = FALSE` in this function because the default value of `flow_mode` has already been set in `stroke()` function and hence duplicated. But I fixed the order of the arguments in `best_link` function.

SarahAlidoost commented:
> Thanks for the speedy implementation, amazing! I have basically only one question/comment - I feel we are very close in getting this working already!

Thanks, I addressed your comments, please let me know if something else needs to be improved.

fnattino reviewed (commented):
Amazing @SarahAlidoost, thanks a lot! 

I have tested it, and it seems to work great. Also tested on the real data set, when this gets added as part of #2 we can add a test that checks this as part of CI (I have raised an issue on this to remember #24).

Just one question: when checking input we now have the following conditions:
```r
  if (attributes && !flow_mode) {
    stop("Stroke attributes can be returned only if `flow_mode = TRUE`)")
  }
```
```r
if (!is.null(from_edge) && (attributes || flow_mode)) {
    stop("from_edge is not compatible with attributes or flow_mode")
```
I remember we discussed the second one together and I think I suggested that `flow_mode` and `from_edge` should be exclusive - but I am now wondering whether this should actually be the case. `flow_mode` only acts on the selection of `best_links` and `from_edge` only acts on `merge_lines`, so maybe one can actually use these together? 

Maybe only `from_edge` and `attribute` should exclude each other, since `from_edge` allows for a segment (or edge) to be reused as part of multiple strokes, so you would not know which stroke label to assign to each edge. For the same reason, `attribute` mode could only be used together with `flow_mode`. 
 
So maybe  we should leave the first condition untouched:
```r
  if (attributes && !flow_mode) {
    stop("Stroke attributes can be returned only if `flow_mode = TRUE`)")
  }
```
but modify the second to allow for `from_edge` and `flow_mode` together? 
```r
if (!is.null(from_edge) && attributes) {
    stop("`from_edge` is not compatible with `attributes = TRUE`")
```
What do you think?

fnattino reviewed (approved):
You are definitely right @SarahAlidoost , `from_edge` will not necessarily "respect" the original edges - so we need to leave the constraints on input as they are. Please go ahead and merge!  :)
